Article about Food Discovery demo (#252)
qdrant/landing_page: qdrant-landing/content/articles/food-discovery-demo.md

however it handles multiple scenarios. Let's dive into them one by one and understand why they are needed.

### Cold start

Recommendation systems struggle with a cold start problem. When a new user joins the system, there is no data about their preferences, so it’s hard to recommend
anything. The same applies to our demo. When you open it, you will see a random selection of dishes, and it changes every time you refresh the page. Internally, 
the demo [chooses some random points](https://github.com/qdrant/demo-food-discovery/blob/6b49e11cfbd6412637d527cdd62fe9b9f74ac699/backend/discovery.py#L70) in the 
vector space.

TODO: add a gif that shows how it works in UI

### Textual search

Since the demo suffers from the cold start problem, we implemented a textual search mode that is useful to start exploring the data. You can type in any text query
by clicking a search icon in the top right corner. The demo will use the CLIP model to encode the query into a vector and then search for the nearest neighbors
in the vector space. 

TODO: add gif showing how it works in UI

> Under the hood this is [a group search query to Qdrant](https://github.com/qdrant/demo-food-discovery/blob/6b49e11cfbd6412637d527cdd62fe9b9f74ac699/backend/discovery.py#L44). 

Author: I tend to use it quite often, but I thought it was a popular term:
https://en.wiktionary.org/wiki/under_the_hood

    ),
    with_vectors=True,
)

# Calculate a mean vector of disliked points
disliked_vectors = np.array([point.vector for point in disliked_points])
mean_vector = np.mean(disliked_vectors, axis=0)
negated_vector = -mean_vector

# Search for nearest neighbors of the negated mean vector
response = client.search_groups(
    settings.QDRANT_COLLECTION,
    query_vector=negated_vector.tolist(),
    group_by=settings.GROUP_BY_FIELD,
    limit=search_query.limit,
)
```

#### Positive and negative feedback

Since the [Recommendation API](https://qdrant.tech/documentation/concepts/search/#recommendation-api) requires at least one positive example, we can use it only when 
the user has liked at least one dish. We could theoretically use the same trick as above and negate the disliked dishes, but it would be a bit weird, as Qdrant has
that feature already built-in, and we can call it just once. It's always better to perform the search server-side. Thus, in this case we just call the Qdrant server
with a list of positive and negative examples, so it can find some points which are close to the positive examples and far from the negative ones.

TODO: code sample

### Location-based search

> Last but not least. Location plays an important role in the food discovery process. You are definitely looking for something you can find nearby, not on the other

Reviewer: ```suggestion
Last but not least, location plays an important role in the food discovery process. You are definitely looking for something you can find nearby, not on the other
```

side of the globe. Therefore, all the search modes accepts your current location as a filtering criteria. This way you can find the best pizza in your neighborhood, 
not in the whole world. Qdrant [geo radius filter](https://qdrant.tech/documentation/concepts/filtering/#geo-radius) is a perfect choice for this. It allows you to
filter the results by distance from a given point. 

```python
from qdrant_client import models

# Create a geo radius filter
query_filter = models.Filter(
    must=[
        models.FieldCondition(
            key="cafe.location",
            geo_radius=models.GeoRadius(
                center=models.GeoPoint(
                    lon=location.longitude,
                    lat=location.latitude,
                ),
                radius=location.radius_km * 1000,
            ),
        )
    ]
)
```

Such a filter needs [a payload index](https://qdrant.tech/documentation/concepts/indexing/#payload-index) to work efficiently, and it was created on a collection

> we used to create the snapshot. So if you import it, the index will be already there.

Reviewer: Cant start the sentences with So, And etc.

---
title: Food Discovery Demo

> short_description: Feeling hungry? Multimodal semantic search makes it easy to find the perfect meal.

Reviewer: ```suggestion
short_description: Feeling hungry? Find the perfect meal with Qdrant's multimodal semantic search.
```

Reviewer: Give me a few more rounds. I really wanna do this well. The demo could be a big hit.

and the demo will update the search results accordingly.

![Recommendation results](/articles_data/food-discovery-demo/recommendation-results.png)

#### Negative feedback only

Qdrant [Recommendation API](https://qdrant.tech/documentation/concepts/search/#recommendation-api) needs at least one positive example to work. However, in our demo
we want to be able to provide only negative examples. This is because we want to be able to say “I don’t like this dish” without having to like anything first.
To achieve this, we use a trick. We negate the vectors of the disliked dishes and use their mean as a query. This way, the disliked dishes will be pushed away 
from the search results. **This works because the cosine distance is based on the angle between two vectors, and the angle between a vector and its negation is 180 degrees.**

![CLIP model](/articles_data/food-discovery-demo/negated-vector.png)

Food Discovery Demo [implements that trick](https://github.com/qdrant/demo-food-discovery/blob/6b49e11cfbd6412637d527cdd62fe9b9f74ac699/backend/discovery.py#L122)

> by calling Qdrant twice. First of all we use the [Scroll API](https://qdrant.tech/documentation/concepts/points/#scroll-points) to find the disliked items, 

Reviewer: ```suggestion
by calling Qdrant twice. Initially, we use the [Scroll API](https://qdrant.tech/documentation/concepts/points/#scroll-points) to find disliked items, 
```

and then calculate a negated mean of all their vectors. That allows using the [Search Groups API](https://qdrant.tech/documentation/concepts/search/#search-groups)

> to find the nearest neighbors of the negated mean vector and we are there!

Reviewer: ```suggestion
to find the nearest neighbors of the negated mean vector.
```

the user has liked at least one dish. We could theoretically use the same trick as above and negate the disliked dishes, but it would be a bit weird, as Qdrant has
that feature already built-in, and we can call it just once to do the job. It's always better to perform the search server-side. Thus, in this case [we just call 
the Qdrant server with a list of positive and negative examples](https://github.com/qdrant/demo-food-discovery/blob/6b49e11cfbd6412637d527cdd62fe9b9f74ac699/backend/discovery.py#L166), 
so it can find some points which are close to the positive examples and far from the negative ones.

```python
response = client.recommend_groups(
    settings.QDRANT_COLLECTION,
    positive=search_query.positive,
    negative=search_query.negative,
    group_by=settings.GROUP_BY_FIELD,
    limit=search_query.limit,
)
```

From the user perspective nothing changes comparing to the previous case.

### Location-based search

Last but not least, location plays an important role in the food discovery process. You are definitely looking for something you can find nearby, not on the other

> side of the globe. Therefore, all the search modes accepts your current location as a filtering criteria, and you can enable it by clicking on “Find near me” icon

Reviewer: ```suggestion
side of the globe. Therefore, your current location can be toggled as a filtering condition. You can enable it by clicking on “Find near me” icon
```

```python
from qdrant_client import models

# Create a geo radius filter
query_filter = models.Filter(
    must=[
        models.FieldCondition(
            key="cafe.location",
            geo_radius=models.GeoRadius(
                center=models.GeoPoint(
                    lon=location.longitude,
                    lat=location.latitude,
                ),
                radius=location.radius_km * 1000,
            ),
        )
    ]
)
```

Such a filter needs [a payload index](https://qdrant.tech/documentation/concepts/indexing/#payload-index) to work efficiently, and it was created on a collection
we used to create the snapshot. When you import it into your instance, the index will be already there.

## Using the demo

The Food Discovery Demo [is available online](https://food-discovery.qdrant.tech/), but if you prefer to run it locally, you can do it with Docker. The 
[README](https://github.com/qdrant/demo-food-discovery/blob/main/README.md) describes all the steps more in detail, but here is a quick start:

```bash
git clone git@github.com:qdrant/demo-food-discovery.git
cd demo-food-discovery
# Create .env file based on .env.example
docker-compose up -d
```

> The demo will be available at `http://localhost:8000`, but you won't be able to search anything until you import the snapshot into your Qdrant instance.

Reviewer: ```suggestion
The demo will be available at `http://localhost:8000`, but you won't be able to search for anything until you import the snapshot into your Qdrant instance.
```

> The demo will be available at `http://localhost:8000`, but you won't be able to search anything until you import the snapshot into your Qdrant instance.

Reviewer: import the snapshot into your Qdrant instance - should link to instructions for how to do this

Author: It has been linked to the documentation.

If you don't want to bother with hosting a local one, you can use the [Qdrant Cloud](https://cloud.qdrant.io/) cluster. 4 GB RAM is enough to load all the 2 million
entries.

## Fork and reuse

The full source code of this demonstration is open-sourced and available for you to fork, adapt, and build upon. Whether you’re looking to understand the mechanics 
of semantic search or to have a foundation to build a larger project, this demo can serve as a starting point. Check out the [Food Discovery Demo repository

> ](https://github.com/qdrant/demo-food-discovery/) to get started. If you have any questions, feel free to reach out to us [on our Discord community](https://qdrant.to/discord).

Reviewer: ```suggestion
](https://github.com/qdrant/demo-food-discovery/) to get started. If you have any questions, feel free to reach out  [through Discord](https://qdrant.to/discord).
```
